exp/ipam/api/v1alpha1/conversion.go

```diff
@@ -49,13 +50,42 @@ func (dst *IPAddressList) ConvertFrom(srcRaw conversion.Hub) error {
 func (src *IPAddressClaim) ConvertTo(dstRaw conversion.Hub) error {
 	dst := dstRaw.(*ipamv1.IPAddressClaim)
 
-	return Convert_v1alpha1_IPAddressClaim_To_v1beta1_IPAddressClaim(src, dst, nil)
+	if err := Convert_v1alpha1_IPAddressClaim_To_v1beta1_IPAddressClaim(src, dst, nil); err != nil {
+		return err
+	}
+
+	if src.ObjectMeta.Labels != nil {
+		dst.Spec.ClusterName = src.ObjectMeta.Labels["cluster.x-k8s.io/cluster-name"]
+		if dst.ObjectMeta.Annotations != nil && dst.ObjectMeta.Annotations["conversion.cluster.x-k8s.io/cluster-name-label-set"] != "" {
+			delete(src.ObjectMeta.Labels, "cluster.x-k8s.io/cluster-name")
+			delete(src.ObjectMeta.Annotations, "conversion.cluster.x-k8s.io/cluster-name-label-set")
+		}
+	}
+
+	return nil
 }
 
 func (dst *IPAddressClaim) ConvertFrom(srcRaw conversion.Hub) error {
 	src := srcRaw.(*ipamv1.IPAddressClaim)
 
-	return Convert_v1beta1_IPAddressClaim_To_v1alpha1_IPAddressClaim(src, dst, nil)
+	if err := Convert_v1beta1_IPAddressClaim_To_v1alpha1_IPAddressClaim(src, dst, nil); err != nil {
+		return err
+	}
+
+	if src.Spec.ClusterName != "" {
+		if dst.ObjectMeta.Labels == nil {
+			dst.ObjectMeta.Labels = map[string]string{}
+		}
+		if _, ok := dst.ObjectMeta.Labels["cluster.x-k8s.io/cluster-name"]; !ok {
+			if dst.ObjectMeta.Annotations == nil {
+				dst.ObjectMeta.Annotations = map[string]string{}
+			}
+			dst.ObjectMeta.Annotations["conversion.cluster.x-k8s.io/cluster-name-label-set"] = "false"
+		}
```

Comment:
This is done to make the fuzzy conversion tests happy. I preferred this over a annotation with a json representation of the old object, but can switch to that "standard" approach as well.

Reply:
I would prefer if we use the standard approach here for consistency

Reply:
Or is it a use case that we intentionally want to cover that controllers written against v1alpha1 might be able to get the cluster-name from the annotation? (which would only work if the spec.clusterName field is set)

Wondering if these controllers can simply upgrade to v1beta1 instead? Or does somebody already rely on the label?

Reply:
I guess if we don't have to support some use cases around this I would prefer to just go with the standard conversion approach and avoid additional complexity.

Reply:
Yeah, the idea was to support old providers. If we don't want that we can also just be forward-compatible instead. I'd be fine with just keeping the annotation when converting old to new, but fuzzy testing is unhappy.

Reply:
Can we keep it this way? Since this is still considered experimental API we can probably remove v1alpha1 rather soon anyway.

Reply:
Discussed this with Jakob here: https://kubernetes.slack.com/archives/C8TSNPY4T/p1714405639788599

Overall we will go the way with the additional annotation. The problem is that the annotation and the funcs we usually use doesn't cover metadata. So we can't use it to backup/restore the cluster-name. But even if we could use it to backup/restore the cluster-name label, we wouldn't want to do it. If someone modifies the cluster-name label on the v1alpha1 type we don't want to preserve the value that that label might have had before the conversion on the v1beta1 types.